### ISSUE TITLE
Validate Dataset assets partially located in both source S3 bucket and Geostore S3 bucket

### ISSUE THREAD

**Issue by imincik** [linz/geostore#1049]
### User Story

So that I can easily update my dataset with small changes, as a data maintainer, I want a way to edit just the assets/metadata I want to update and submit them to back to the Geostore to update an existing dataset.

Instead of, having to copy the full dataset somewhere else, do the edits and then submit the full dataset back to the geostore to create a new version.

#### Acceptance Criteria

- [ ] Given a source bucket directory with a partial dataset, when validating the dataset, then geostore will validate the files in the Geostore bucket for each file, if they are not present in the source bucket.
- [ ] Given a source bucket directory with a partial dataset, when the dataset is validated successfully, then the geostore will copy all files from the source bucket to the geostore bucket
- [ ] Given all files are valid, when validation step is complete, then the files from the geostore bucket are *not* recopied into new versions of themselves (do we need to test this?)
- [ ] Given a source bucket directory with no files that could be part of a new version, when validating the dataset (ideally the first step), then the process stops and a message is returned to the user

#### Additional context

Suggested design:

Creating the dataset and the first version works exactly like now. The interesting stuff happens when someone wants to change only a subset of metadata and/or assets:

1. The user downloads the files they want to change. (This means when they want to change an asset they also need to download the metadata file with the relevant checksum.)
2. They make their changes in their own bucket.
3. They create a new version by specifying the ID of the dataset and the link to the dataset directory in their own bucket (depends on #1237 to be implemented)
4. The magic happens because of two changes in the Geostore:
   - During validation, if the Geostore doesn't find a file in the source bucket, it instead reads the same file from the Geostore bucket. This way, the last version of the dataset in the Geostore *fills the holes* in the source bucket during validation.
   - Only the files in the source bucket are copied across to the Geostore.

This means that:

- The user only has to copy across the files which they want to change.
- If they didn't copy across the top-level metadata file (because they didn't want to modify it) they have to _counter-intuitively_ submit a URL to a non-existing file.
- If they copy across files which they *don't* modify before submitting, the files will still be copied across, so it's up to the user to make sure the source bucket only contains changed files.
- We don't delete files which are no longer referenced.

Bonus features:

- ~~We could make the import process more intuitive by forcing the top-level metadata file to always be `collection.json`, so the user just has to submit the URL to the dataset _folder_ when creating a version. That way they don't need to refer to a possibly non-existing file in the source bucket.~~ See #1237 
- ~~Alternatively we could determine the root metadata file by the user specifying only the dataset ID. Unfortunately this would mean more complexity in the API, since there would be two different ways to create a new version.~~
- ~~Determine which assets are no longer referenced and delete them. Lower priority since dataset maintainers probably rarely do this?~~ See #1224 
- ~~Determine which metadata files are no longer referenced and delete them. Even lower priority since metadata files are generally small?~~ See #1224 
- Avoid copying unchanged files from the source bucket. This would ease the burden on the user, since they won't need to be careful about their source bucket containing only changed files. It would involve comparing two large lists of files in memory, which might mean bumping the memory use of the relevant Lambda a fair bit.
- Skip checksum/STAC validation when reading a file from the Geostore. This would speed up the import of a small set of changes considerably.

#### Tasks

- [x] Work out 'detailed' design in a workshop - split to new task 3 estimate
- [ ] Add possibility to validate Dataset assets partially located in both source S3 bucket and Geostore S3 bucket (latest version).
- [ ] Update documentation on how to use this funcitonality

#### Definition of Ready

- [ ] This story is **ready** to work on
  - [ ] Independent (story is independent of all other tasks)
  - [ ] Negotiable (team can decide how to design and implement)
  - [ ] Valuable (from a user perspective)
  - [ ] Estimate value applied (agreed by team)
  - [ ] Small (so as to fit within an iteration)
  - [ ] Testable (in principle, even if there isn't a test for it yet)
  - [ ] Environments are ready to meet definition of done
  - [ ] Resources required to implement will be ready
  - [ ] Everyone understands and agrees with the tasks to complete the story
  - [ ] Release value (e.g. Iteration 3) applied
  - [ ] Sprint value (e.g. Aug 1 - Aug 15) applied

#### Definition of Done

- [ ] This story is **done**:
  - [ ] Acceptance criteria completed
  - [ ] Automated tests are passing
  - [ ] Code is peer reviewed and pushed to master
  - [ ] Deployed successfully to test environment
  - [ ] Checked against [CODING guidelines](https://github.com/linz/geostore/blob/master/CODING.md)
  - [ ] Relevant new tasks are added to backlog and communicated to the team
  - [ ] Important decisions recorded in the issue ticket
  - [ ] Readme/Changelog/Diagrams are updated
  - [ ] Product Owner has approved acceptance criteria as complete
  - [ ] Meets non-functional requirements:
    - [ ] Scalability (data): Can scale to 300TB of data and 100,000,000 files and ability to
          increase 10% every year
    - [ ] Scability (users): Can scale to 100 concurrent users
    - [ ] Cost: Data can be stored at < 0.5 NZD per GB per year
    - [ ] Performance: A large dataset (500 GB and 50,000 files - e.g. Akl aerial imagery) can be
          validated, imported and stored within 24 hours
    - [ ] Accessibility: Can be used from LINZ networks and the public internet
    - [ ] Availability: System available 24 hours a day and 7 days a week, this does not include
          maintenance windows < 4 hours and does not include operational support
    - [ ] Recoverability: RPO of fully imported datasets < 4 hours, RTO of a single 3 TB dataset <
          12 hours

**Comment by billgeo** [comment]
HI @imincik tried rewriting this as a user story with acc. criteria (added your sentence as a task) Am I understanding this right?

**Comment by billgeo** [comment]
Discussion with Victor

Could be around an 8 estimate, but some uncertainty, may require more team discussion

**Comment by billgeo** [comment]
Confirmed this is working as expected. Have raised on bug found during testing, https://github.com/linz/geostore/issues/1738
but we can closet his issue.
